File: server/services/compareEngine.ts
import geoData from '../data/geo/nationality-regions.json';
import positionGroups from '../data/position-groups.json';
import { CompareResult, Theme } from '../types';

type CountryInfo = { continent: string; groups: string[] };

const countries = geoData.countries as Record<string, CountryInfo>;
const proximityGroups = geoData.proximityGroups as string[][];

function normalizeStr(val: unknown): string {
  if (val === null || val === undefined) return '';
  return String(val).trim().toLowerCase();
}

function compareExact(a: unknown, b: unknown): CompareResult {
  if (normalizeStr(a) === normalizeStr(b)) return 'hit';
  return 'miss';
}

function compareNumber(
  guess: unknown,
  answer: unknown,
  closeThreshold: number,
  absoluteClose?: number
): CompareResult {
  const g = Number(guess);
  const a = Number(answer);
  if (isNaN(g) || isNaN(a)) return 'miss';
  if (g === a) return 'hit';
  if (absoluteClose !== undefined && Math.abs(g - a) <= absoluteClose) {
    return 'close';
  }
  if (a !== 0 && Math.abs(g - a) / Math.abs(a) <= closeThreshold) {
    return 'close';
  }
  if (a === 0 && Math.abs(g) <= (absoluteClose ?? 1)) return 'close';
  return 'miss';
}

function positionTokens(position: string): string[] {
  const raw = String(position ?? '').trim();
  if (!raw) return [];
  if (raw.includes('/')) {
    return raw
      .split('/')
      .map((part) => part.trim())
      .filter(Boolean);
  }
  return [raw];
}

function getPositionGroup(theme: Theme, position: string): string | null {
  const groups = positionGroups[theme as keyof typeof positionGroups] as Record<
    string,
    string[]
  >;
  if (!groups) return null;
  const norm = normalizeStr(position);
  for (const [group, members] of Object.entries(groups)) {
    if (members.some((m) => normalizeStr(m) === norm)) return group;
  }
  for (const token of positionTokens(position)) {
    const tokenNorm = normalizeStr(token);
    for (const [group, members] of Object.entries(groups)) {
      if (members.some((m) => normalizeStr(m) === tokenNorm)) return group;
    }
  }
  return null;
}

function comparePosition(theme: Theme, guess: unknown, answer: unknown): CompareResult {
  const gFull = normalizeStr(String(guess ?? ''));
  const aFull = normalizeStr(String(answer ?? ''));
  if (gFull === aFull) return 'hit';

  const gTokens = positionTokens(String(guess ?? '')).map(normalizeStr);
  const aTokens = positionTokens(String(answer ?? '')).map(normalizeStr);
  if (gTokens.some((gt) => aTokens.includes(gt))) return 'hit';

  const gGroup = getPositionGroup(theme, String(guess ?? ''));
  const aGroup = getPositionGroup(theme, String(answer ?? ''));
  if (gGroup && aGroup && gGroup === aGroup) return 'close';
  return 'miss';
}

function compareNationality(guess: unknown, answer: unknown): CompareResult {
  const g = String(guess ?? '').trim();
  const a = String(answer ?? '').trim();
  if (g === a) return 'hit';

  const gInfo = countries[g];
  const aInfo = countries[a];
  if (!gInfo || !aInfo) return 'miss';

  // 仅按更细的 groups（北欧、西欧、东欧、独联体、中东、东南亚等）判断接近，不再使用大洲级别匹配
  for (const group of proximityGroups) {
    if (group.includes(g) && group.includes(a)) return 'close';
  }

  const gGroups = new Set(gInfo.groups);
  for (const grp of aInfo.groups) {
    if (gGroups.has(grp)) return 'close';
  }

  return 'miss';
}

function compareNullable(guess: unknown, answer: unknown): CompareResult {
  const gNull = guess === null || guess === undefined || guess === '';
  const aNull = answer === null || answer === undefined || answer === '';
  if (gNull && aNull) return 'hit';
  if (gNull || aNull) return 'miss';
  return compareExact(guess, answer);
}

export function parseDraftYear(draft: unknown): number | null {
  const m = String(draft ?? '').match(/(\d{4})年/);
  return m ? Number(m[1]) : null;
}

function parseDraftRound(draft: unknown): number | null {
  const m = String(draft ?? '').match(/(\d{4})年第(\d+)轮/);
  return m ? Number(m[2]) : null;
}

function isUndrafted(draft: unknown): boolean {
  return /落选秀|undrafted/i.test(String(draft ?? ''));
}

function compareDraft(guess: unknown, answer: unknown): CompareResult {
  const g = String(guess ?? '');
  const a = String(answer ?? '');
  if (g === a) return 'hit';

  const gUndrafted = isUndrafted(g);
  const aUndrafted = isUndrafted(a);
  if (gUndrafted && aUndrafted) return 'hit';
  if (gUndrafted || aUndrafted) return 'miss';

  const gy = parseDraftYear(g);
  const ay = parseDraftYear(a);
  if (gy == null || ay == null) return compareExact(guess, answer);
  if (gy === ay) {
    const gr = parseDraftRound(g);
    const ar = parseDraftRound(a);
    if (gr != null && ar != null && gr === ar) return 'hit';
    return 'close';
  }
  return 'miss';
}

/** Answer draft year vs guess: later = answer drafted after guess year */
export function getDraftYearDirection(
  guess: unknown,
  answer: unknown
): 'later' | 'earlier' | null {
  if (isUndrafted(guess) || isUndrafted(answer)) return null;
  const gy = parseDraftYear(guess);
  const ay = parseDraftYear(answer);
  if (gy == null || ay == null || gy === ay) return null;
  return ay > gy ? 'later' : 'earlier';
}

/** UI hint text for NBA draft field comparisons */
export function getDraftCompareHint(
  result: CompareResult,
  guess: unknown,
  answer: unknown
): string | null {
  if (result === 'hit') return null;
  if (result === 'close') return '轮次不对';
  const dir = getDraftYearDirection(guess, answer);
  const year = parseDraftYear(guess);
  if (!dir || year == null) return null;
  return dir === 'later' ? `晚于${year}` : `早于${year}`;
}

const NUMERIC_RULES: Record<
  Theme,
  Record<string, { threshold?: number; absolute?: number }>
> = {
  csgo: {
    age: { absolute: 2 },
    rating: { absolute: 0.08 },
    top20Count: { absolute: 1 },
  },
  football: {
    age: { absolute: 2 },
    marketValue: { threshold: 0.2 },
    height: { absolute: 3 },
  },
  nba: {
    age: { absolute: 2 },
    height: { absolute: 3 },
    playoffCount: { absolute: 1 },
  },
  anime: {
    age: { absolute: 5 },
    height: { absolute: 10 },
    powerLevel: { absolute: 8 },
  },
  pokemon: {
    baseStatTotal: { absolute: 30 },
    hp: { absolute: 15 },
    attack: { absolute: 15 },
    defense: { absolute: 15 },
    spAttack: { absolute: 15 },
    spDefense: { absolute: 15 },
    speed: { absolute: 15 },
  },
};

const POSITION_FIELDS = new Set(['position']);
const NATIONALITY_FIELDS = new Set(['nationality', 'nationalTeam']);
const NULLABLE_FIELDS = new Set(['club', 'school', 'type2']);

const POKEMON_STAT_FIELDS = new Set([
  'baseStatTotal',
  'hp',
  'attack',
  'defense',
  'spAttack',
  'spDefense',
  'speed',
]);

function parseEggGroups(val: unknown): string[] {
  return String(val ?? '')
    .split('、')
    .map((s) => s.trim())
    .filter(Boolean);
}

/** Single vs dual egg group: shared group but different count → close with answer hint. */
export function compareEggGroup(
  guessValue: unknown,
  answerValue: unknown
): { result: CompareResult; hint?: string } {
  const g = parseEggGroups(guessValue);
  const a = parseEggGroups(answerValue);
  if (g.length === 0 || a.length === 0) return { result: 'miss' };

  const gKey = [...g].sort().join('|');
  const aKey = [...a].sort().join('|');
  if (gKey === aKey) return { result: 'hit' };

  const shared = g.filter((x) => a.includes(x));
  if (shared.length > 0 && g.length !== a.length) {
    return {
      result: 'close',
      hint: `答案为${a.length >= 2 ? '双蛋群' : '单蛋群'}`,
    };
  }
  return { result: 'miss' };
}

export function compareField(
  theme: Theme,
  field: string,
  guessValue: unknown,
  answerValue: unknown
): CompareResult {
  if (field === 'name') {
    return compareExact(guessValue, answerValue);
  }

  if (NULLABLE_FIELDS.has(field)) {
    return compareNullable(guessValue, answerValue);
  }

  if (NATIONALITY_FIELDS.has(field)) {
    return compareNationality(guessValue, answerValue);
  }

  if (POSITION_FIELDS.has(field)) {
    return comparePosition(theme, guessValue, answerValue);
  }

  if (theme === 'nba' && field === 'draft') {
    return compareDraft(guessValue, answerValue);
  }

  if (theme === 'pokemon' && field === 'eggGroup') {
    return compareEggGroup(guessValue, answerValue).result;
  }

  const rule = NUMERIC_RULES[theme]?.[field];
  if (rule) {
    // Handle string ages like "千年" for anime
    if (typeof guessValue === 'string' || typeof answerValue === 'string') {
      return compareExact(guessValue, answerValue);
    }
    return compareNumber(
      guessValue,
      answerValue,
      rule.threshold ?? 0,
      rule.absolute
    );
  }

  // string enum fields: team, anime, genre, race, occupation
  const exact = compareExact(guessValue, answerValue);
  if (exact === 'hit') return 'hit';

  // race/occupation same category close for anime
  if (theme === 'anime' && ['race', 'occupation'].includes(field)) {
    const gGroup = getPositionGroup('anime', String(guessValue));
    const aGroup = getPositionGroup('anime', String(answerValue));
    if (gGroup && aGroup && gGroup === aGroup) return 'close';
  }

  // team close: same first word or substring (weaker)
  if (field === 'team' || field === 'anime' || field === 'club') {
    const g = normalizeStr(guessValue);
    const a = normalizeStr(answerValue);
    if (g.includes(a) || a.includes(g)) return 'close';
  }

  return 'miss';
}

/** When numeric guess misses, indicate higher or lower than answer */
export function getNumericDirection(
  theme: Theme,
  field: string,
  guessValue: unknown,
  answerValue: unknown
): 'higher' | 'lower' | null {
  const rule = NUMERIC_RULES[theme]?.[field];
  if (!rule) return null;
  if (typeof guessValue === 'string' || typeof answerValue === 'string') return null;

  const g = Number(guessValue);
  const a = Number(answerValue);
  if (isNaN(g) || isNaN(a) || g === a) return null;
  return g > a ? 'higher' : 'lower';
}

export function formatValue(val: unknown, field?: string): string | number | null {
  const empty =
    val === null || val === undefined || (typeof val === 'string' && val.trim() === '');
  if (field === 'club' && empty) return '无';
  if (field === 'school' && empty) return '无';
  if (field === 'type2' && empty) return '无';
  if (empty) return null;
  if (typeof val === 'boolean') return val ? '会' : '不会';
  if (typeof val === 'number') return val;
  return String(val);
}

export function isPokemonStatField(field: string): boolean {
  return POKEMON_STAT_FIELDS.has(field);
}

File: server-go/internal/services/compare_engine.go
package services

import (
	"fmt"
	"regexp"
	"sort"
	"strings"

	"guess-who/server-go/internal/types"
)

// CompareResult 对应 TS 类型：hit | close | miss
type CompareResult string

const (
	Hit   CompareResult = "hit"
	Close CompareResult = "close"
	Miss  CompareResult = "miss"
)

type numericRule struct {
	threshold *float64
	absolute  *float64
}

var numericRules = map[types.Theme]map[string]numericRule{
	types.ThemeCSGO: {
		"age":        {absolute: ptrFloat(2)},
		"rating":     {absolute: ptrFloat(0.08)},
		"top20Count": {absolute: ptrFloat(1)},
	},
	types.ThemeFootball: {
		"age":         {absolute: ptrFloat(2)},
		"marketValue": {threshold: ptrFloat(0.2)},
		"height":      {absolute: ptrFloat(3)},
	},
	types.ThemeNBA: {
		"age":          {absolute: ptrFloat(2)},
		"height":       {absolute: ptrFloat(3)},
		"playoffCount": {absolute: ptrFloat(1)},
	},
	types.ThemePokemon: {
		"baseStatTotal": {absolute: ptrFloat(30)},
		"hp":            {absolute: ptrFloat(15)},
		"attack":        {absolute: ptrFloat(15)},
		"defense":       {absolute: ptrFloat(15)},
		"spAttack":      {absolute: ptrFloat(15)},
		"spDefense":     {absolute: ptrFloat(15)},
		"speed":         {absolute: ptrFloat(15)},
	},
}

var (
	nullableFields    = map[string]bool{"club": true, "school": true, "type2": true}
	nationalityFields = map[string]bool{"nationality": true, "nationalTeam": true}
	positionFields    = map[string]bool{"position": true}
	pokemonStatFields = map[string]bool{
		"baseStatTotal": true, "hp": true, "attack": true, "defense": true,
		"spAttack": true, "spDefense": true, "speed": true,
	}
)

var (
	draftYearRe  = regexp.MustCompile(`(\d{4})年`)
	draftRoundRe = regexp.MustCompile(`(\d{4})年第(\d+)轮`)
)

func ptrFloat(f float64) *float64 { return &f }

func normalizeStr(v any) string {
	if v == nil {
		return ""
	}
	return strings.ToLower(strings.TrimSpace(fmt.Sprintf("%v", v)))
}

func toFloat64(v any) (float64, bool) {
	switch val := v.(type) {
	case float64:
		return val, true
	case int:
		return float64(val), true
	case int64:
		return float64(val), true
	case string:
		var f float64
		if _, err := fmt.Sscanf(strings.TrimSpace(val), "%f", &f); err != nil {
			return 0, false
		}
		return f, true
	default:
		return 0, false
	}
}

func toFloat(v any) float64 {
	f, _ := toFloat64(v)
	return f
}

func compareExact(a, b any) CompareResult {
	if normalizeStr(a) == normalizeStr(b) {
		return Hit
	}
	return Miss
}

func compareNumber(guess, answer any, closeThreshold float64, absoluteClose *float64) CompareResult {
	g, gOk := toFloat64(guess)
	a, aOk := toFloat64(answer)
	if !gOk || !aOk {
		return Miss
	}
	if g == a {
		return Hit
	}
	if absoluteClose != nil && absFloat(g-a) <= *absoluteClose {
		return Close
	}
	if a != 0 && absFloat(g-a)/absFloat(a) <= closeThreshold {
		return Close
	}
	absFallback := 1.0
	if absoluteClose != nil {
		absFallback = *absoluteClose
	}
	if a == 0 && absFloat(g) <= absFallback {
		return Close
	}
	return Miss
}

func absFloat(x float64) float64 {
	if x < 0 {
		return -x
	}
	return x
}

func positionTokens(pos string) []string {
	raw := strings.TrimSpace(pos)
	if raw == "" {
		return nil
	}
	if strings.Contains(raw, "/") {
		parts := strings.Split(raw, "/")
		out := make([]string, 0, len(parts))
		for _, p := range parts {
			p = strings.TrimSpace(p)
			if p != "" {
				out = append(out, p)
			}
		}
		return out
	}
	return []string{raw}
}

func getPositionGroup(theme types.Theme, position string) *string {
	groups := GetPositionGroups(theme)
	if len(groups) == 0 {
		return nil
	}
	norm := normalizeStr(position)
	for group, members := range groups {
		for _, m := range members {
			if normalizeStr(m) == norm {
				g := group
				return &g
			}
		}
	}
	for _, token := range positionTokens(position) {
		tokenNorm := normalizeStr(token)
		for group, members := range groups {
			for _, m := range members {
				if normalizeStr(m) == tokenNorm {
					g := group
					return &g
				}
			}
		}
	}
	return nil
}

func comparePosition(theme types.Theme, guess, answer any) CompareResult {
	gFull := normalizeStr(fmt.Sprintf("%v", guess))
	aFull := normalizeStr(fmt.Sprintf("%v", answer))
	if gFull == aFull {
		return Hit
	}
	gTokens := make([]string, 0)
	for _, t := range positionTokens(fmt.Sprintf("%v", guess)) {
		gTokens = append(gTokens, normalizeStr(t))
	}
	aTokens := make([]string, 0)
	for _, t := range positionTokens(fmt.Sprintf("%v", answer)) {
		aTokens = append(aTokens, normalizeStr(t))
	}
	for _, gt := range gTokens {
		for _, at := range aTokens {
			if gt == at {
				return Hit
			}
		}
	}
	gGroup := getPositionGroup(theme, fmt.Sprintf("%v", guess))
	aGroup := getPositionGroup(theme, fmt.Sprintf("%v", answer))
	if gGroup != nil && aGroup != nil && *gGroup == *aGroup {
		return Close
	}
	return Miss
}

func compareNullable(guess, answer any) CompareResult {
	gNull := isEmptyValue(guess)
	aNull := isEmptyValue(answer)
	if gNull && aNull {
		return Hit
	}
	if gNull || aNull {
		return Miss
	}
	return compareExact(guess, answer)
}

func isEmptyValue(v any) bool {
	if v == nil {
		return true
	}
	if s, ok := v.(string); ok {
		return strings.TrimSpace(s) == ""
	}
	return false
}

// ParseDraftYear 从 NBA 选秀字符串提取选秀年份。
func ParseDraftYear(draft any) *int {
	s := fmt.Sprintf("%v", draft)
	m := draftYearRe.FindStringSubmatch(s)
	if len(m) == 2 {
		y := 0
		fmt.Sscanf(m[1], "%d", &y)
		return &y
	}
	return nil
}

func parseDraftRound(draft any) *int {
	s := fmt.Sprintf("%v", draft)
	m := draftRoundRe.FindStringSubmatch(s)
	if len(m) == 3 {
		r := 0
		fmt.Sscanf(m[2], "%d", &r)
		return &r
	}
	return nil
}

func isUndrafted(draft any) bool {
	s := strings.ToLower(fmt.Sprintf("%v", draft))
	return strings.Contains(s, "落选秀") || strings.Contains(s, "undrafted")
}

func compareDraft(guess, answer any) CompareResult {
	g := fmt.Sprintf("%v", guess)
	a := fmt.Sprintf("%v", answer)
	if g == a {
		return Hit
	}
	gu := isUndrafted(g)
	au := isUndrafted(a)
	if gu && au {
		return Hit
	}
	if gu || au {
		return Miss
	}
	gy := ParseDraftYear(g)
	ay := ParseDraftYear(a)
	if gy == nil || ay == nil {
		return compareExact(guess, answer)
	}
	if *gy == *ay {
		gr := parseDraftRound(g)
		ar := parseDraftRound(a)
		if gr != nil && ar != nil && *gr == *ar {
			return Hit
		}
		return Close
	}
	return Miss
}

// GetDraftYearDirection 当答案与猜测的选秀年份不同时返回 later/earlier。
func GetDraftYearDirection(guess, answer any) *string {
	if isUndrafted(guess) || isUndrafted(answer) {
		return nil
	}
	gy := ParseDraftYear(guess)
	ay := ParseDraftYear(answer)
	if gy == nil || ay == nil || *gy == *ay {
		return nil
	}
	if *ay > *gy {
		s := "later"
		return &s
	}
	s := "earlier"
	return &s
}

// GetDraftCompareHint 返回 NBA 选秀对比的 UI 提示文本。
func GetDraftCompareHint(result CompareResult, guess, answer any) *string {
	if result == Hit {
		return nil
	}
	if result == Close {
		s := "轮次不对"
		return &s
	}
	dir := GetDraftYearDirection(guess, answer)
	year := ParseDraftYear(guess)
	if dir == nil || year == nil {
		return nil
	}
	var s string
	if *dir == "later" {
		s = fmt.Sprintf("晚于%d", *year)
	} else {
		s = fmt.Sprintf("早于%d", *year)
	}
	return &s
}

func parseEggGroups(val any) []string {
	parts := strings.Split(fmt.Sprintf("%v", val), "、")
	out := make([]string, 0, len(parts))
	for _, s := range parts {
		s = strings.TrimSpace(s)
		if s != "" && s != "<nil>" {
			out = append(out, s)
		}
	}
	return out
}

// EggGroupCompare 保存生蛋群对比结果及可选提示。
type EggGroupCompare struct {
	Result CompareResult
	Hint   *string
}

// CompareEggGroup 比较宝可梦生蛋群，接近未命中时给出单/双生蛋群提示。
func CompareEggGroup(guessValue, answerValue any) EggGroupCompare {
	g := parseEggGroups(guessValue)
	a := parseEggGroups(answerValue)
	if len(g) == 0 || len(a) == 0 {
		return EggGroupCompare{Result: Miss}
	}
	gKey := strings.Join(sortedCopy(g), "|")
	aKey := strings.Join(sortedCopy(a), "|")
	if gKey == aKey {
		return EggGroupCompare{Result: Hit}
	}
	shared := 0
	for _, x := range g {
		for _, y := range a {
			if x == y {
				shared++
				break
			}
		}
	}
	if shared > 0 && len(g) != len(a) {
		hint := "答案为单蛋群"
		if len(a) >= 2 {
			hint = "答案为双蛋群"
		}
		return EggGroupCompare{Result: Close, Hint: &hint}
	}
	return EggGroupCompare{Result: Miss}
}

func sortedCopy(s []string) []string {
	out := append([]string(nil), s...)
	sort.Strings(out)
	return out
}

// CompareField 比较猜测与答案在某一字段上的值。
func CompareField(theme types.Theme, field string, guessValue, answerValue any) CompareResult {
	if field == "name" {
		return compareExact(guessValue, answerValue)
	}
	if nullableFields[field] {
		return compareNullable(guessValue, answerValue)
	}
	if nationalityFields[field] {
		return CompareNationality(guessValue, answerValue)
	}
	if positionFields[field] {
		return comparePosition(theme, guessValue, answerValue)
	}
	if theme == types.ThemeNBA && field == "draft" {
		return compareDraft(guessValue, answerValue)
	}
	if theme == types.ThemePokemon && field == "eggGroup" {
		return CompareEggGroup(guessValue, answerValue).Result
	}

	if rule, ok := numericRules[theme][field]; ok {
		if _, isStr := guessValue.(string); isStr {
			return compareExact(guessValue, answerValue)
		}
		if _, isStr := answerValue.(string); isStr {
			return compareExact(guessValue, answerValue)
		}
		threshold := 0.0
		if rule.threshold != nil {
			threshold = *rule.threshold
		}
		return compareNumber(guessValue, answerValue, threshold, rule.absolute)
	}

	exact := compareExact(guessValue, answerValue)
	if exact == Hit {
		return Hit
	}
	if field == "team" || field == "club" {
		g := normalizeStr(guessValue)
		a := normalizeStr(answerValue)
		if strings.Contains(g, a) || strings.Contains(a, g) {
			return Close
		}
	}
	return Miss
}

// GetNumericDirection 数值猜测未命中时返回 higher/lower。
func GetNumericDirection(theme types.Theme, field string, guessValue, answerValue any) *string {
	rule, ok := numericRules[theme][field]
	if !ok {
		return nil
	}
	_ = rule
	if _, isStr := guessValue.(string); isStr {
		return nil
	}
	if _, isStr := answerValue.(string); isStr {
		return nil
	}
	g, gOk := toFloat64(guessValue)
	a, aOk := toFloat64(answerValue)
	if !gOk || !aOk || g == a {
		return nil
	}
	var s string
	if g > a {
		s = "higher"
	} else {
		s = "lower"
	}
	return &s
}

// FormatValue 格式化字段值以供展示。
func FormatValue(val any, field ...string) any {
	f := ""
	if len(field) > 0 {
		f = field[0]
	}
	empty := val == nil
	if s, ok := val.(string); ok {
		empty = strings.TrimSpace(s) == ""
	}
	if f == "club" && empty {
		return "无"
	}
	if f == "school" && empty {
		return "无"
	}
	if f == "type2" && empty {
		return "无"
	}
	if empty {
		return nil
	}
	if b, ok := val.(bool); ok {
		if b {
			return "会"
		}
		return "不会"
	}
	switch v := val.(type) {
	case float64:
		if v == float64(int64(v)) {
			return int(v)
		}
		return v
	case int, int64:
		return v
	default:
		return fmt.Sprintf("%v", val)
	}
}

// IsPokemonStatField 判断字段是否为宝可梦种族值字段。
func IsPokemonStatField(field string) bool {
	return pokemonStatFields[field]
}

func guessKnowsMove(entry types.CharacterEntry, move string) bool {
	raw, ok := entry["gen3LevelMoves"]
	if !ok {
		return false
	}
	switch moves := raw.(type) {
	case []string:
		for _, m := range moves {
			if m == move {
				return true
			}
		}
	case []interface{}:
		for _, item := range moves {
			if s, ok := item.(string); ok && s == move {
				return true
			}
		}
	}
	return false
}

// CompareAllFields 比较猜测与答案的全部活跃字段。
func CompareAllFields(
	theme types.Theme,
	guess types.CharacterEntry,
	answer types.CharacterEntry,
	activeFields []string,
	compareMove *string,
) []types.FieldCompare {
	fields := GetThemeFields(theme, activeFields)
	out := make([]types.FieldCompare, 0, len(fields))

	for _, fd := range fields {
		field := fd.Field
		label := fd.Label

		if theme == types.ThemePokemon && field == "learnableMove" {
			knows := false
			if compareMove != nil && *compareMove != "" {
				knows = guessKnowsMove(guess, *compareMove)
			}
			var guessVal any
			if knows {
				guessVal = FormatValue("会")
			} else {
				guessVal = FormatValue("不会")
			}
			answerVal := FormatValue("会")
			result := Miss
			if knows {
				result = Hit
			}
			moveLabel := label
			if compareMove != nil && *compareMove != "" {
				moveLabel = "可学习：" + *compareMove
			}
			var ansOut any
			showAnswer := result == Hit
			if showAnswer {
				ansOut = answerVal
			}
			out = append(out, types.FieldCompare{
				Field: field, Label: moveLabel,
				GuessValue: guessVal, AnswerValue: ansOut,
				Result: string(result), ShowAnswer: showAnswer,
			})
			continue
		}

		var guessCompare, answerCompare any
		if field == "name" {
			guessCompare = GetNameFieldValue(guess, theme)
			answerCompare = GetNameFieldValue(answer, theme)
		} else {
			guessCompare = GetCharacterField(guess, field, theme)
			answerCompare = GetCharacterField(answer, field, theme)
		}

		guessDisplay := guessCompare
		answerDisplay := answerCompare

		if theme == types.ThemeNBA && field == "team" {
			guessDisplay = GetNBATeamDisplay(fmt.Sprintf("%v", guessCompare))
			answerDisplay = GetNBATeamDisplay(fmt.Sprintf("%v", answerCompare))
		}
		if theme == types.ThemeNBA && field == "position" {
			gPos := GetNBAPositionDisplay(fmt.Sprintf("%v", guessCompare))
			aPos := GetNBAPositionDisplay(fmt.Sprintf("%v", answerCompare))
			guessDisplay = gPos.Zh
			answerDisplay = aPos.En
		}

		compareGuess := guessDisplay
		compareAnswer := answerDisplay
		if theme == types.ThemeNBA && field == "position" {
			compareGuess = guessCompare
			compareAnswer = answerCompare
		}

		if theme == types.ThemePokemon && field == "eggGroup" {
			egg := CompareEggGroup(compareGuess, compareAnswer)
			gv := FormatValue(guessDisplay, field)
			av := FormatValue(answerDisplay, field)
			var ansOut any
			if egg.Result == Hit {
				ansOut = av
			}
			out = append(out, types.FieldCompare{
				Field: field, Label: label,
				GuessValue: gv, AnswerValue: ansOut,
				Result: string(egg.Result), ShowAnswer: egg.Result == Hit,
				Hint: egg.Hint,
			})
			continue
		}

		gv := FormatValue(guessDisplay, field)
		av := FormatValue(answerDisplay, field)
		result := CompareField(theme, field, compareGuess, compareAnswer)

		var direction *string
		if result != Hit {
			if theme == types.ThemeNBA && field == "draft" {
				direction = GetDraftYearDirection(compareGuess, compareAnswer)
			} else {
				direction = GetNumericDirection(theme, field, compareGuess, compareAnswer)
			}
		}

		var hint *string
		if theme == types.ThemeNBA && field == "draft" && result != Hit {
			hint = GetDraftCompareHint(result, compareGuess, compareAnswer)
		}

		var ansOut any = av
		if field == "position" && result != Hit {
			ansOut = nil
		}

		out = append(out, types.FieldCompare{
			Field: field, Label: label,
			GuessValue: gv, AnswerValue: ansOut,
			Result: string(result), ShowAnswer: result == Hit,
			Direction: direction, Hint: hint,
		})
	}
	return out
}
